### PR TITLE
Integrity check for maven, npm and pypi 

### PR DESCRIPTION
**Pull request by mehab**
Addresses [issue](https://github.com/DependencyTrack/hyades/issues/699)

The integrity check functionality (maven, npm and pypi). 
Hyades-apiserver PR: https://github.com/DependencyTrack/hyades-apiserver/pull/261

### CODE REVIEW

File: proto/src/main/proto/org/hyades/repometaanalysis/v1/repo_meta_analysis.proto [hunk 1]
```diff
@@ -27,11 +27,37 @@ message AnalysisResult {
...
+  // The Component for which the integrity check was done
+  Component component = 1;
+  //md5 hash match status of component
+  HashMatchStatus md5_hash_match = 2;
+  //sha1 hash match status of component
+  HashMatchStatus sha1_hash_match = 3;
+  //sha256 hash match status of component
+  HashMatchStatus sha256_hash_match = 4;
+  //Identifier of the repository with which the hashes were compared
+  string repository = 5;
```
Comment (nscuro) [inline proto/src/main/proto/org/hyades/repometaanalysis/v1/repo_meta_analysis.proto:40]: The information currently being sent here is the repository URL (e.g. `https://repo1.maven.org/maven2/`). Is this intended? If so, both the field name and description should reflect that.

File: repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+            try {
+                final var analyzerComponent = new org.hyades.persistence.model.Component();
+                analyzerComponent.setPurl(component.getPurl());
+                analyzerComponent.setMd5(component.getMd5Hash());
+                analyzerComponent.setSha1(component.getSha1Hash());
+                analyzerComponent.setSha256(component.getSha256Hash());
+                UUID uuid = UUID.fromString(component.getUuid());
+                analyzerComponent.setUuid(uuid);
+                analyzerComponent.setId((long) component.getComponentId());
+                try (CloseableHttpResponse response = analyzer.getIntegrityCheckResponse(new PackageURL(component.getPurl()))) {
```
Comment (nscuro) [inline repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java:182]: `getIntegrityCheckResponse` can return `null` in some cases. Should we handle that case explicitly, and do we really want to cache `null`?

File: repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        //
+        // NOTE: The result of this query can potentially be cached for at least a few minutes.
+        // Executing it for every single component feels excessive.
+        // Quarkus has Hibernate L2 cache enabled by default, we just need to opt in to using
+        // it for this query: https://quarkus.io/guides/hibernate-orm#caching-of-queries
+        // Should be tested whether throughput can be improved this way.
+        //changed this to joinexisting because with new transaction, it is not fetching values that were inserted from
+        // existing previous transaction and returning empty result
+        return QuarkusTransaction.joiningExisting()
+                .call(() -> repoEntityRepository.findEnabledRepositoriesByType(repositoryType));
```
[the remark this answers is not in the document]
Comment (sahibamittal) [reply repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java:119]: integrity_check_enabled is being checked for just after this method call : 
`for (Repository repository : getApplicableRepositories(analyzer.supportedRepositoryType())) {
            if (repository.isIntegrityCheckEnabled()) {`

File: repository-meta-analyzer/src/main/java/org/hyades/model/IntegrityModel.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import org.hyades.proto.repointegrityanalysis.v1.HashMatchStatus;
+
+import java.io.Serializable;
+import java.util.Date;
+
+public class IntegrityModel implements Serializable {
+
+    private Component component;
+    private Date publishedTimestamp;
+    private HashMatchStatus md5HashMatched;
```
Comment (VithikaS) [inline repository-meta-analyzer/src/main/java/org/hyades/model/IntegrityModel.java:13]: Very minor: variables like md5HashMatchStatus would be more readable here. 
Method name isMd5HashMatched looks like should be returning boolean. getMd5HashMatchStatus might be more readable

File: repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        }
+        // Produce "empty" result in case no repository did yield a satisfactory result.
+        context().forward(inputRecord
+                .withValue(IntegrityResult.newBuilder().setComponent(component)
+                        .setMd5HashMatch(HashMatchStatus.HASH_MATCH_STATUS_UNKNOWN).setSha1HashMatch(HashMatchStatus.HASH_MATCH_STATUS_UNKNOWN)
+                        .setSha256HashMatch(HashMatchStatus.HASH_MATCH_STATUS_UNKNOWN).build())
+                .withTimestamp(context().currentSystemTimeMs()));
+    }
+
+    private PackageURL mustParsePurl(final String purl) {
```
Comment (VithikaS) [inline repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java:94]: Repeated code in both MetaAnalyzer and IntegrityAnalyzer

File: repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            return new PackageURL(purl);
+        } catch (MalformedPackageURLException e) {
+            throw new IllegalStateException("""
+                    The provided PURL is invalid, even though it should have been
+                    validated in a previous processing step
+                    """, e);
+        }
+    }
+
+    private List<Repository> getApplicableRepositories(final RepositoryType repositoryType) {
```
Comment (VithikaS) [inline repository-meta-analyzer/src/main/java/org/hyades/processor/IntegrityAnalyzerProcessor.java:105]: This also looks repeated from MetaAnalyzer

**Comment by mehab** [comment]
Closing this pull request after reconsidering the design as per requirement of having published date for newly fetched components all the time. Adding the detailed meeting notes on issue https://github.com/DependencyTrack/hyades/issues/699

**Pull request closed without merging**